### PR TITLE
KAFKA-15073: Add a Github action to mark PRs as stale

### PR DESCRIPTION
**Pull request by mumrah**
This patch uses a Github workflow to mark inactive PRs with the `stale` label. The workflow uses the ["stale" action](https://github.com/actions/stale) which has many features including a number of filtering options and the ability to auto-closing issues.

For this patch, I've disable auto-closing PRs and just have it add the label for PRs with more than 90 days of inactivity. 

I've also only included a manual trigger. We can eventually include a cron schedule for automatic nightly running of the workflow.

I have tested this out on my fork of Kafka and it worked as expected: https://github.com/mumrah/kafka/actions/workflows/stale.yml

Here you can see when it actually marked an old PR as stale https://github.com/mumrah/kafka/actions/runs/5203928815/jobs/9387453473

image

### CODE REVIEW

File: .github/workflows/stale.yml
Comment (jlprat) [inline .github/workflows/stale.yml:47]: I would probably add the message there:
`stale-pr-message: 'This PR is stale because it has been open 90 days with no activity.'`

**Review by jlprat — COMMENTED** [review]
I think this looks good, but I still have some open questions about how we can effectively work with this in these 2 particular cases:
- Author needs to perform some actions
- Reviewer is needed

I looked at the documentation of the action and I don't seem to see how we could separate this is 2 runs, we would like to flag differently PRs with no activity for 90 days depending on:
- Changes Requested / Merge Conflict --> Stale and message stating that please address the issues
- No reviews, or only accepted reviews and no merge conflicts --> We could have a "Maintainer-needed' label

Unfortunately, I don't see how we can do this with this action easily

**Comment by mumrah** [comment]
Thanks for taking a look @jlprat! 

For the more complex workflow, we could use https://github.com/actions/github-script. This basically lets you do anything that the Github API allows (kind of like using Groovy in a Gradle build 😄).

I also found https://github.com/marketplace/actions/auto-label-merge-conflicts which would add a label to PRs with conflicts. This could be used to create two separate workflows of the "stale" action. 

I think there's a lot of flexibility with the actions available.

Do you agree that we can scope this PR to "marking old PRs as stale" for the sake of clearing our backlog?

**Comment by jlprat** [comment]
> I also found https://github.com/marketplace/actions/auto-label-merge-conflicts which would add a label to PRs with conflicts. This could be used to create two separate workflows of the "stale" action.

This is a good one, I think we can combine these 2 actions and make them run to accomplish what we need. Mark PRs with merge conflicts with a label, and then we can run the stale job twice, one excluding the conflict label and another one only for the conflict label.

> Do you agree that we can scope this PR to "marking old PRs as stale" for the sake of clearing our backlog?

Yes this makes sense. I'll work on the other one.

File: .github/workflows/stale.yml
Comment (jlprat) [inline .github/workflows/stale.yml:45]: I think 90 days is a long time but looking at our queue, I think it's a realistic value to start with.

**Review by jlprat — APPROVED** [review]
The flow looks good to me and I think we could try it out.

File: .github/workflows/stale.yml
Comment (danielgospodinow) [inline .github/workflows/stale.yml:18]: Maybe we can add an additional comment here to explain that there's a plan to use this cron schedule for nightly running of the workflow.

**Review by danielgospodinow — COMMENTED** [review]
IMO, this is a great idea and initiative. 😄

File: .github/workflows/stale.yml
Comment (dajac) [inline .github/workflows/stale.yml]: Should we expand the comment a little more? For instance, we could explain we the contributors could do (e.g. nag us, fix conflicts, close, etc)? We could also apologize for the delay…

Comment (dajac) [reply .github/workflows/stale.yml]: For instance, i like this one: https://github.com/apache/spark/blob/master/.github/workflows/stale.yml

File: .github/workflows/stale.yml
Comment (dajac) [reply .github/workflows/stale.yml:18]: Any reason why we don’t enable it?

File: .github/workflows/stale.yml
Comment (dajac) [inline .github/workflows/stale.yml:46]: Why don’t we close them as well after another 30 days? It won’t be close of a reviewer or the contributor reacts. I am saying this because without closing them, it does not really help with reducing the backlog.

File: .github/workflows/stale.yml
Comment (mumrah) [reply .github/workflows/stale.yml:18]: We can't actually test the functionality of the action in a PR, so I was going to commit it without the automatic schedule so we could manually run it as a test

File: .github/workflows/stale.yml
Comment (mumrah) [reply .github/workflows/stale.yml:46]: I'd like to in a future PR. Since we can't test the action until it's committed, I wanted to hold off on doing anything "destructive" like closing a bunch of PRs :) 

After testing this manually with the dry-run option, I'll raise a PR to add the schedule and close PRs

Comment (jlprat) [reply .github/workflows/stale.yml:46]: I also think that having a dry run first and see that the right PRs would be flagged is a good approach

File: .github/workflows/stale.yml
Comment (jlprat) [inline .github/workflows/stale.yml]: There are plenty of PRs that are stale but no further action from the author is needed. I would add a conditional in there.

"...please update it with the latest from trunk if there is a merge conflict and ask..."

**Review by jlprat — APPROVED** [review]
LGTM. Thanks @mumrah

**Pull request merged**
